### PR TITLE
PYTHON-1552 Adds test that an exception thrown while writing a GridFS file ensures chunks are saved, but not associated with a file.

### PR DESCRIPTION
**Pull request by bcwarner**
The [spec for GridFS](https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#id21) says not to rollback any changes that occur as a result of errors while uploading a file, so this test was added to check that the state doesn't get rolled back.

### CODE REVIEW

File: test/test_grid_file.py [hunk 1]
```diff
@@ -675,6 +675,23 @@ def test_context_manager(self):
         with GridOut(self.db.fs, infile._id) as outfile:
             self.assertEqual(contents, outfile.read())
 
+    def test_exception_file_non_existence(self):
+        contents = b"Imagine this is some important data..."
+
+        try:
```
Comment (ShaneHarvey) [inline test/test_grid_file.py:681]: Instead of try/except it would be cleaner to use `self.assertRaises(CorruptGridFile):` https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises

File: test/test_grid_file.py [hunk 2]
```diff
@@ -675,6 +675,23 @@ def test_context_manager(self):
...
+            with GridIn(self.db.fs, filename="important") as infile:
+                infile.write(contents)
+                raise CorruptGridFile("Test exception")
```
Comment (ShaneHarvey) [inline test/test_grid_file.py:684]: CorruptGridFile would never actually be raised by GridIn.write so it probably makes more sense to use pymongo.errors.ConnectionFailure or something.

File: test/test_grid_file.py [hunk 3]
```diff
@@ -675,6 +675,23 @@ def test_context_manager(self):
...
+                # Never calls infile.flush.
```
Comment (ShaneHarvey) [inline test/test_grid_file.py:685]: Is it important to call (or not call) flush here?

File: test/test_grid_file.py [hunk 4]
```diff
@@ -675,6 +675,23 @@ def test_context_manager(self):
...
+        except CorruptGridFile as e:
+            pass
+
+        # Expectation: File chunks are written, entry in files doesn't appear.
+        with GridOut(self.db.fs, infile._id) as outfile:
+            self.assertEqual(contents, outfile.read())
+
+        self.assertIsNone(self.db.files.find_one({"_id": infile._id}))
```
Comment (ShaneHarvey) [inline test/test_grid_file.py:693]: This should be `self.db.fs.files.find_one`

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Nice work so far. The test is essentially what we want but now we need to actually fix the bug and update the test to prove the bug is fixed.

File: gridfs/__init__.py [hunk 5]
```diff
@@ -135,10 +135,8 @@ def put(self, data: Any, **kwargs: Any) -> Any:
            w=0 writes to GridFS are now prohibited.
         """
         grid_file = GridIn(self.__collection, **kwargs)
-        try:
-            grid_file.write(data)
-        finally:
-            grid_file.close()
+        grid_file.write(data)
+        grid_file.close()
```
Comment (ShaneHarvey) [inline gridfs/__init__.py:139]: Let's use a with-statement here instead of calling close explicitly.

File: test/test_grid_file.py (hunk 3 above)
Comment (bcwarner) [reply test/test_grid_file.py]: Turns out flush is internally handled

File: gridfs/grid_file.py [hunk 6]
```diff
@@ -396,9 +396,10 @@ def __enter__(self) -> "GridIn":
     def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> Any:
         """Support for the context manager protocol.
 
-        Close the file and allow exceptions to propagate.
+        Close the file if no exceptions occur and allow exceptions to propagate.
         """
-        self.close()
+        if exc_type is None:
+            self.close()
```
Comment (ShaneHarvey) [inline gridfs/grid_file.py:402]: I think we should still mark the GridIn object as closed. Otherwise calling close explicitly after the with-statement would end up saving the file incorrectly:
```python
with GridIn(...) as file:
    file.write(...)
    raise Exception()
file.close() # This should be a no-op because the file is already marked as "closed"
```

Can you update the test to cover this case as well?

File: test/test_gridfs.py [hunk 7]
```diff
@@ -540,7 +540,7 @@ def test_gridfs_secondary_lazy(self):
 
         # Connects, doesn't create index.
         self.assertRaises(NoFile, fs.get_last_version)
-        self.assertRaises(NotPrimaryError, fs.put, "data")
+        self.assertRaises(NotPrimaryError, fs.put, "data", encoding="utf-8")
```
Comment (ShaneHarvey) [inline test/test_gridfs.py:543]: Why was this needed?

**Comment by bcwarner** [comment]
Changed `__exit__` so that it only calls `close()` when no exceptions are thrown. Before this it would always call close() and create the associated doc in `files`. I also fixed a unit test where a call to `GridFS.put` didn't provide an encoding and failed.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
There are a few more places in GridFS+the docs that still use try/finally. Can you update those to use with-statements too?

File: test/test_gridfs.py (hunk 7 above)
Comment (bcwarner) [reply test/test_gridfs.py:543]: I think the try-finally statement originally in `GridFS.put` suppressed an error that should have shown up without the `encoding` kwarg there. `GridFS.put` relies on `GridIn.write`, the latter of which requires that an encoding is set for `str` objects, otherwise it will raise an exception.

File: gridfs/__init__.py [hunk 8]
```diff
@@ -109,11 +109,8 @@ def put(self, data: Any, **kwargs: Any) -> Any:
 
         Equivalent to doing::
 
-          try:
-              f = new_file(**kwargs)
+          with GridIn(self.database[collection], **kwargs) as f:
```
Comment (ShaneHarvey) [inline gridfs/__init__.py:112]: Th9is example should still use new_file, actually `fs.new_file`.

File: test/test_gridfs.py (hunk 7 above)
Comment (ShaneHarvey) [reply test/test_gridfs.py:543]: Can you try removing this change and share the traceback of the test failure (if it fails)?

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: test/test_gridfs.py (hunk 7 above)
Comment (bcwarner) [reply test/test_gridfs.py:543]: Yeah, this is the one test that fails:

```
======================================================================
ERROR: test_gridfs_secondary_lazy (test.test_gridfs.TestGridfsReplicaSet)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/benwarner/Documents/pymongo/gridfs/grid_file.py", line 350, in write
    read = data.read
AttributeError: 'str' object has no attribute 'read'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/benwarner/Documents/pymongo/gridfs/grid_file.py", line 357, in write
    data = data.encode(self.encoding)
  File "/Users/benwarner/Documents/pymongo/gridfs/grid_file.py", line 252, in __getattr__
    raise AttributeError("GridIn object has no attribute '%s'" % name)
AttributeError: GridIn object has no attribute 'encoding'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/benwarner/Documents/pymongo/test/test_gridfs.py", line 543, in test_gridfs_secondary_lazy
    self.assertRaises(NotPrimaryError, fs.put, "data")
  File "/Users/benwarner/opt/anaconda3/lib/python3.9/unittest/case.py", line 739, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "/Users/benwarner/opt/anaconda3/lib/python3.9/unittest/case.py", line 201, in handle
    callable_obj(*args, **kwargs)
  File "/Users/benwarner/Documents/pymongo/gridfs/__init__.py", line 136, in put
    grid_file.write(data)
  File "/Users/benwarner/Documents/pymongo/gridfs/grid_file.py", line 359, in write
    raise TypeError("must specify an encoding for file in order to write str")
TypeError: must specify an encoding for file in order to write str
```

Comment (ShaneHarvey) [reply test/test_gridfs.py:543]: Oooh I get it now.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Great stuff! I just did a final check for places that should be using a with statement and found these examples that should be updated:
```
doc/migrate-to-pymongo4.rst:885:  grid_in.write(file_data)
gridfs/__init__.py:528:          grid_in.write("data I want to store!")
gridfs/__init__.py:586:          grid_in.write("data I want to store!")
```

**Review by ShaneHarvey — APPROVED** [review]
When merging please update the commit message to something a little more succinct like:
`PYTHON-1552 Avoid uploading partial/corrupt GridFS file after an error occurs`

If you want to add more info you can add it to the body of the commit message:
```
PYTHON-1552 Avoid uploading partial/corrupt GridFS file after an error occurs

If an exception is thrown while...
```
